sec_certs/dataset/dataset.py

```diff
@@ -1,73 +1,92 @@
 from sec_certs.config.configuration import config
 from sec_certs.dataset.cpe import CPEDataset
 from sec_certs.dataset.cve import CVEDataset
 from sec_certs.model.cpe_matching import CPEClassifier
 from sec_certs.sample.certificate import Certificate
 from sec_certs.sample.cpe import CPE
-from sec_certs.serialization.json import ComplexSerializableType, serialize
+from sec_certs.serialization.json import ComplexSerializableType, get_class_fullname, serialize
+from sec_certs.utils.tqdm import tqdm
 
 logger = logging.getLogger(__name__)
 
-T = TypeVar("T")
+
+@dataclass
+class AuxillaryDatasets:
+    cpe_dset: CPEDataset | None = None
+    cve_dset: CVEDataset | None = None
+
+
 CertSubType = TypeVar("CertSubType", bound=Certificate)
+AuxillaryDatasetsSubType = TypeVar("AuxillaryDatasetsSubType", bound=AuxillaryDatasets)
 DatasetSubType = TypeVar("DatasetSubType", bound="Dataset")
 
 
-class Dataset(Generic[CertSubType], ABC):
+class Dataset(Generic[CertSubType, AuxillaryDatasetsSubType], ComplexSerializableType, ABC):
+    @dataclass
+    class DatasetInternalState(ComplexSerializableType):
+        meta_sources_parsed: bool = False
+        artifacts_downloaded: bool = False
+        pdfs_converted: bool = False
+        auxillary_datasets_processed: bool = False
+        certs_analyzed: bool = False
+
+        def __bool__(self):
+            return any(vars(self))
+
     def __init__(
         self,
-        certs: Dict[str, CertSubType] = dict(),
-        root_dir: Optional[Path] = None,
-        name: str = "dataset name",
-        description: str = "dataset_description",
+        certs: dict[str, CertSubType] = dict(),
+        root_dir: str | Path = constants.DUMMY_NONEXISTING_PATH,
+        name: str | None = None,
+        description: str = None,
+        state: DatasetInternalState | None = None,
+        auxillary_datasets: AuxillaryDatasetsSubType | None = None,
     ):
-        if not root_dir:
-            root_dir = Path.cwd() / (type(self).__name__).lower()
-        self._root_dir = root_dir
+        self.certs = certs
+
         self.timestamp = datetime.now()
         self.sha256_digest = "not implemented"
-        self.name = name
-        self.description = description
-        self.certs = certs
+        self.name = name if name else type(self).__name__.lower() + "_dataset"
+        self.description = description if description else "No description provided"
+        self.state = state if state else self.DatasetInternalState()
+
+        if not auxillary_datasets:
+            self.auxillary_datasets = AuxillaryDatasets()
+        else:
+            self.auxillary_datasets = auxillary_datasets
+
+        self.root_dir = Path(root_dir)
 
     @property
     def root_dir(self) -> Path:
         return self._root_dir
 
     @root_dir.setter
-    def root_dir(self, new_dir: Union[str, Path]) -> None:
+    def root_dir(self: DatasetSubType, new_dir: str | Path) -> None:
```

Comment:
> Should this be a public setter and not a function like _set_root_dir? It allows for getting the dataset into a state where its root_dir points to wherever, which might not even exist. Thus I think this should be more private and only used internally. But maybe we want users to be able to set this.

well if we allow the users to instantiate dataset without a `root_dir`, they need to set it afterwards somehow. So it must be public. The design right now is that you can use the dataset as a datastructure without any explicit paths. As soon as you want to process it somehow, it will scream that you must set `root_dir`. 

> Another thing is that this tests whether the root_dir is a file but not whether it is actually a directory. I understand that we may not want to check that, as it may not (yet) be a directory when this gets called but still, checking that it is not a file seems extra and a weird check. But this is probably just a matter of expectations. I would not expect such a check here.

I don't mind if it doesn't exist. I just mind if it's a file. Then creating a directory on the same path would overwrite it.

Reply:
I see, makes sense.